cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/RegularExpression.java

```diff
@@ -20,34 +23,55 @@
     public RegularExpression(Pattern expressionRegexp, ParameterTypeRegistry parameterTypeRegistry) {
         this.expressionRegexp = expressionRegexp;
         this.parameterTypeRegistry = parameterTypeRegistry;
-        treeRegexp = new TreeRegexp(expressionRegexp);
+        this.treeRegexp = new TreeRegexp(expressionRegexp);
     }
 
     @Override
-    public List<Argument<?>> match(String text) {
-        List<ParameterType<?>> parameterTypes = new ArrayList<>();
+    public List<Argument<?>> match(String text, Type... typeHints) {
+        final ObjectMapper objectMapper = parameterTypeRegistry.getObjectMapper();
+        final List<ParameterType<?>> parameterTypes = new ArrayList<>();
+        int typeHintIndex = 0;
         for (GroupBuilder groupBuilder : treeRegexp.getGroupBuilder().getChildren()) {
-            String parameterTypeRegexp = groupBuilder.getSource();
+            final String parameterTypeRegexp = groupBuilder.getSource();
+            final Type typeHint = typeHintIndex < typeHints.length ? typeHints[typeHintIndex++] : Object.class;
 
             ParameterType<?> parameterType = parameterTypeRegistry.lookupByRegexp(parameterTypeRegexp, expressionRegexp, text);
-            if (parameterType == null) parameterType = new ParameterType<>(
-                    null,
-                    parameterTypeRegexp,
-                    String.class,
-                    new Transformer<String>() {
-                        @Override
-                        public String transform(String arg) {
-                            return arg;
-                        }
-                    }
-            );
+
+            if (parameterType != null && !parameterType.isAnonymous()) {
+                // Use hint for validation only
```

Comment:
On second thought I think we should leave the validation out. The language specific step definitions have much more type information available. There is just no way we can get this right for every language here.

Reply:
I think we should do validation when we *do* have the information (i.e when the tyoe is a class). We’ll be able to make slightly better error messages because we have more context. The drawback is that we won’t always be able to (with generic types) but I think that’s ok.